### PR TITLE
nspawn: check if kernel supports userns as early as possible

### PR DESCRIPTION
**Pull request by tixxdz**
If the kernel do not support user namespace then one of the children
created by nspawn parent will fail at clone(CLONE_NEWUSER) with a generic
error EINVAL and without logging the error. At the same time the parent
may also try to setup the user namespace and will fail with another
error.

To improve this, check if the kernel supports user namespace as early as
possible and while we are it rename the function determine_names() to
determine_container_environment(). This way we combine the checks into
the same function.

### CODE REVIEW

**Comment by teg** [comment]
Not sure if reporposing determine_names() rather than adding a new function is the right thing to do, but apart from that (and my inline nitpick) this looks good to me.

**Comment by tixxdz** [comment]
@teg ok thanks for the hints. Will wait for another comment for determine_names(), update to log_error_errno() and open a new PR.

**Comment by poettering** [comment]
I think this check should be at the end of parse_argv() or so...

File: src/nspawn/nspawn.c [hunk 1]
```diff
@@ -3948,7 +3948,13 @@ static int on_orderly_shutdown(sd_event_source *s, const struct signalfd_siginfo
         return 0;
 }
 
-static int determine_names(void) {
+/*
+ * nspawn uses multiple kernel and other features, some of them
+ * can be checked in their appropriate functions when we setup the
+ * feature, but there are others features that must be chcked as
```
Comment (jonboulle) [inline src/nspawn/nspawn.c:3954]: checked

File: src/nspawn/nspawn.c [hunk 2]
```diff
@@ -3948,7 +3948,13 @@ static int on_orderly_shutdown(sd_event_source *s, const struct signalfd_siginfo
...
+ * early as possible in order to avoid bad or generic error codes,
```
Comment (jonboulle) [inline src/nspawn/nspawn.c:3955]: trailing comma

**Comment by tixxdz** [comment]
Hi all, thank you for the comments, the new PR is here https://github.com/systemd/systemd/pull/231

So closing this one!

**Pull request closed without merging**